Tasks/Common/ios-signing-common/ios-signing-common.ts

```diff
@@ -602,8 +598,83 @@ export async function getP12CommonName(p12Path: string, p12Pwd: string) {
     return commonName;
 }
 
-export async function deleteCert(keychainPath: string, certSha1Hash: string) {
+export async function deleteCert(keychainPath: string, certSha1Hash: string): Promise<void> {
     let deleteCert: ToolRunner = tl.tool(tl.which('security', true));
     deleteCert.arg(['delete-certificate', '-Z', certSha1Hash, keychainPath]);
     await deleteCert.exec();
 }
+
+export async function getP12PrivateKeyName(p12Path: string, p12Pwd: string): Promise<string> {
+    //openssl pkcs12 -in <p12Path> -nocerts -passin pass:"<p12Pwd>" -passout pass:"<p12Pwd>" | grep 'friendlyName'
+    //We pipe through grep so we we don't log the private key, even encrypted.
+    tl.debug('getting the P12 private key name');
+    let opensslPath: string = tl.which('openssl', true);
+    let openssl: ToolRunner = tl.tool(opensslPath);
+    if (!p12Pwd) {
+        // if password is null or not defined, set it to empty
+        p12Pwd = '';
+    }
+    openssl.arg(['pkcs12', '-in', p12Path, '-nocerts', '-passin', 'pass:' + p12Pwd, '-passout', 'pass:' + p12Pwd]);
+
+    let grepPath: string = tl.which('grep', true);
+    let grep: ToolRunner = tl.tool(grepPath);
+    grep.arg(['friendlyName']);
+    openssl.pipeExecOutputToTool(grep);
+
+    let privateKeyName: string;
+    openssl.on('stdline', function (line: string) {
+        if (line) {
+            const match = line.match(/friendlyName: (.*)/);
+            if (match && match[1]) {
+                privateKeyName = match[1].trim();
+            }
+        }
+    });
+
+    await openssl.exec();
+    tl.debug('P12 private key name = ' + privateKeyName);
+    if (!privateKeyName) {
+        throw tl.loc('P12PrivateKeyNameNotFound', p12Path);
+    }
+
+    return privateKeyName;
+}
+
+/**
+ * Set the partition_id ACL so codesign has permission to use the signing key.
+ */
+async function setKeyPartitionList(keychainPath: string, keychainPwd: string, privateKeyName: string) {
```

Comment:
Fastlane's solution if you're curious:

https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/keychain_importer.rb